### PR TITLE
logging: use journald logging for all systemd services

### PR DESCRIPTION
**Pull request by orndorffgrant**
## Why is this needed?

This PR brings journald logging to all systemd services, plus a few small related refactors.

## Test Steps

```
./tools/test-in-lxd.sh xenial
apt update
# ensure logs of esm_cache end up in journald
```

## Checklist

 - [x] I have updated or added any unit tests accordingly
 - [ ] I have updated or added any integration tests accordingly
 - [n/a] Changes here need to be documented, and this was done in: 

## Does this PR require extra reviews?

 - [ ] Yes
 - [x] No

### CODE REVIEW

**Review by aciba90 — APPROVED** [review]
LGTM, thanks for the refactors.

**Comment by lucasmoura** [comment]
@orndorffgrant should we use our JSON format when logging to `journalctl` ? I am just wondering if we need to be consistent with the other logs there. But maybe this is not needed

**Comment by orndorffgrant** [comment]
@lucasmoura I think it should use the json logging to be consistent. The `setup_journald_logging` function does use the JsonArrayFormatter and I just updated the integration test to actually check the json output. Does that align with what you were thinking?

**Comment by lucasmoura** [comment]
My rationale is that it shouldn't to be honest, just  to align with the other ` journalctl` entries. But I don't think is an actual problem if we keep the JSON output

**Comment by orndorffgrant** [comment]
@sergiodj @athos-ribeiro @lucasmoura After our discussion in Riga, I've updated this to remove the json formatting in journald logs.

We still want to include multiple fields with structure, though, so I've introduced a simpler (but more fragile) `|`-delimited format.

Taking the example from my last comment here, it would now look like this:
```
...
"_CMDLINE":"/usr/bin/python3 /usr/lib/ubuntu-advantage/esm_cache.py",
"__CURSOR":"s=e4ea46b1825d4b5d833f9f8a48dccd3c;i=862;b=28db989000224dc39694c489bad64176;m=40a4ff859d;t=608138598a742;x=35915d735a50dc59",
"MESSAGE":"WARNING|ubuntupro.apt|fail|872|Failed to fetch ESM Apt Cache item: https://esm.ubuntu.com/apps/ubuntu/dists/xenial-apps-security/InRelease|{}",
"SYSLOG_IDENTIFIER":"python3"
...
```

Do you think this is better or worse than the json formatting?

**Comment by athos-ribeiro** [comment]
> @sergiodj @athos-ribeiro @lucasmoura After our discussion in Riga, I've updated this to remove the json formatting in journald logs.
> 
> We still want to include multiple fields with structure, though, so I've introduced a simpler (but more fragile) `|`-delimited format.
> 
> Taking the example from my last comment here, it would now look like this:
> 
> ```
> ...
> "_CMDLINE":"/usr/bin/python3 /usr/lib/ubuntu-advantage/esm_cache.py",
> "__CURSOR":"s=e4ea46b1825d4b5d833f9f8a48dccd3c;i=862;b=28db989000224dc39694c489bad64176;m=40a4ff859d;t=608138598a742;x=35915d735a50dc59",
> "MESSAGE":"WARNING|ubuntupro.apt|fail|872|Failed to fetch ESM Apt Cache item: https://esm.ubuntu.com/apps/ubuntu/dists/xenial-apps-security/InRelease|{}",
> "SYSLOG_IDENTIFIER":"python3"
> ...
> ```
> 
> Do you think this is better or worse than the json formatting?

Thanks, Grant!

Would you mind showing what an old (current) log entry would look like compared to the new format?
Do you think there are people out there parsing those logs somehow and would this break their workflows?

**Comment by orndorffgrant** [comment]
Thanks for looking @athos-ribeiro!

Prior to this PR, most services logged to /var/log/ubuntu-advantage.log in the JSONArray format. This was not done in a multi-process-safe way. Any messages that ended up in journald were accidental and did not contain any structure, e.g. it would just be the message string like "Failed to fetch ESM Apt Cache item: https://esm.ubuntu.com/apps/ubuntu/dists/xenial-apps-security/InRelease".

Two services already logged to journald in the JSONArray format.

So there are several changes here from the perspective of someone interested in our logs:
1. Some logs that used to appear in ubuntu-advantage.log will no longer appear
2. More logs will show up in journald for their respective services
3. logs in journald that used to be JSONArray formatted will no longer be JSONArray formatted
4. logs in journald that used to be just a string will now have some structure and extra information 

We haven't considered backwards compatibility of our logs very seriously in the past. The biggest breaking change, though, would be moving away from JSONArray formatting for the services that were already using it. Anything else would've been handled as a string, so adding structure wouldn't break it significantly IMO.

**Comment by panlinux** [comment]
> Thanks for looking @athos-ribeiro!
> 
> Prior to this PR, most services logged to /var/log/ubuntu-advantage.log in the JSONArray format. This was not done in a multi-process-safe way. Any messages that ended up in journald were accidental and did not contain any structure, e.g. it would just be the message string like "Failed to fetch ESM Apt Cache item: https://esm.ubuntu.com/apps/ubuntu/dists/xenial-apps-security/InRelease".
> 
> Two services already logged to journald in the JSONArray format.
> 
> So there are several changes here from the perspective of someone interested in our logs:
> 
> 1. Some logs that used to appear in ubuntu-advantage.log will no longer appear
> 2. More logs will show up in journald for their respective services
> 3. logs in journald that used to be JSONArray formatted will no longer be JSONArray formatted
> 4. logs in journald that used to be just a string will now have some structure and extra information
> 
> We haven't considered backwards compatibility of our logs very seriously in the past. The biggest breaking change, though, would be moving away from JSONArray formatting for the services that were already using it. Anything else would've been handled as a string, so adding structure wouldn't break it significantly IMO.

This is quite the breaking change. If I, as a user, would be parsing the logs, I would be very annoyed with such a change in an LTS upgrade.
I Have you considered a config option to toggle between these logging formats? On upgrade, the logging format wouldn't change, but on a fresh install, it would be what this PR is proposing?

**Comment by panlinux** [comment]
Other points:
1. You should document the logging. How it's done, in which format, where.
2. Are there more logging changes coming? When will this be done?

File: features/unattached_commands.feature [hunk 1]
```diff
@@ -383,19 +383,18 @@ Feature: Command behaviour when unattached
...
         """
-        raise FetchFailedException\(\)
+        raise FetchFailedException()
         """
-        When I run `ls /var/crash/` with sudo
-        Then stdout does not match regexp:
+        Then stdout matches regexp:
         """
-        _usr_lib_ubuntu-advantage_esm_cache
+        WARNING|ubuntupro.apt|fail|\d+|Failed to fetch ESM Apt Cache item: https://esm.ubuntu.com/apps/ubuntu/dists/<release>-apps-security/InRelease|{}
```
Comment (lucasmoura) [inline features/unattached_commands.feature:392]: Just to double check, why is the function name `fail` here ?

Comment (orndorffgrant) [reply features/unattached_commands.feature:392]: Inside `update_esm_caches`, an `EsmAcquireProgress` class is defined with a `fail` method that is called by the `apt` library when an error occurs. That method is where we log this line

**Comment by orndorffgrant** [comment]
@panlinux @lucasmoura I've updated this PR so it only does the following:

- Removes not-multiprocess-safe logging to same file from multiple processes
- Each systemd service that previously did not log to journald now does
  - These services log in JSON format

There are no breaking changes.

Is this okay?

**Review by renanrodrigo — APPROVED** [review]
imho this is a good changeset after all the discussion

File: uaclient/cli/__init__.py [hunk 2]
```diff
@@ -886,6 +888,7 @@ def action_disable(args, *, cfg, **kwargs):
     return 0 if ret else 1
 
 
+<<<<<<< HEAD
```
Comment (renanrodrigo) [inline uaclient/cli/__init__.py:891]: oh no

Comment (orndorffgrant) [reply uaclient/cli/__init__.py:891]: ugh

Comment (orndorffgrant) [reply uaclient/cli/__init__.py]: fixed

**Comment by orndorffgrant** [comment]
CI fail unrelated

**Comment by orndorffgrant** [comment]
I'm going ahead and merging the least controversial version of this PR (new logging to journald using JSON format).
We can revisit at SRU review time if necessary.

**Pull request merged**